### ISSUE TITLE
A role doesn't have access to attribute objects of a role it "consumes"

### ISSUE THREAD

**Issue by vrurg** [rakudo/rakudo#4306]
## The Problem

It is normal for a class to have direct access to attribute objects of roles it consumes:

```raku
role R { has $.attr = 42; }
class C does R {
    method attr {
        "attr:" ~ $!attr
    } 
}
```

But same doesn't apply to roles:

```raku
role R1 {
    has $.attr = 42;
}
role R2 does R1 {
    method attr {
        "attr:" ~ $!attr
    }
}
class C does R2 { 
}
```

The above fails to compile with:

> ```
> Attribute $!attr not declared in role R2
> ```

**Comment by jnthn** [comment]
This cannot be provided in general, because until the time we compose a role into a class, we don't reliably know the role we are composing (due to, for example, parameters and the fact that many roles may share the same shortname).

**Comment by vrurg** [comment]
The problem doesn't have other workaround but via `nqp::getattr`:

```raku
method attr {
    nqp::getattr(nqp::decont(self), ::?CLASS, '$!attr')
}
```

Or via much slower:

```raku
    method attr {
        self.^get_attribute_for_usage('$!attr').get_value(self);
    }
```

The problem can be resolved if `Metamodel::ParametricRoleHOW` gets a version of `get_attribute_for_usage` method which would traverse consumed roles.

**Comment by vrurg** [comment]
@jnthn I have to go now, but I understand that full solution would require lazy resolution and should be postponed until RakuAST is ready. But for simple cases my proposal above could work for start.

**Comment by jnthn** [comment]
> should be postponed until RakuAST is ready

I don't think it really provides a solution in this case. The roles may be from different compilation units, and the AST is long gone at that point.

**Comment by vrurg** [comment]
@jnthn a spesh plugin?

**Comment by vrurg** [comment]
A spesh plugin what just a quick though when I already was on the run. But basically the idea reflects what I consider a possible solution for the problem.

First of all, the only reason for me to refer to RakuAST is that current implementation would require some significant amount of patching of Grammar/Actions/World and I'm not sure it makes sense to waste time on this.

What should be helpful here is the new dispatching unless I expect of it more that it will be able to do. So, for roles instead of producing code directly accessing the attribute, a dispatcher could be used which will resolve the attribute at runtime.

**Comment by lizmat** [comment]
Why don't we codegen unknown attributes in a role with:
```
nqp::getattr(self, ::?CLASS, '$!attr')
```
and add a worry about it?
